lib/final-api/model_extensions.rb

```diff
@@ -3,86 +3,138 @@
-      operator = 'IN'
+
+      array = query.scan(expression_pattern)
+      return [['name', ':', query]] if array.empty?
+      wrong_keys = []
+      result = array.map do |item|
+        query_key = item[0].downcase
+        operator = item[1]
+        query_value = item[2]
+        wrong_keys << query_key if query_value.to_s.empty?
+
+        column = SEARCH_TOKENS_DEF.select { |k| k.include? query_key }.values.first
+        wrong_keys << query_key if column.nil?
+
+        [
+          column,
+          operator,
+          query_value.tr("\"", '')
+        ]
+      end
+
+      raise InvalidQueryError,
+            "Wrong search definition(s) specified: #{wrong_keys.join(', ')}" unless wrong_keys.empty?
+
+      result
     end
 
-    case operator
-    when '='
-      # makes search quite slow (and without index)
-      # but users are unaware of types, and could write: WHERE id = 'string'
-      # which leads to error:
-      #   PG::InvalidTextRepresentation: ERROR:  invalid input syntax for integer
-      column = "(#{column})::text"
-    when ':'
-      operator = 'ILIKE'
-      term = "%#{term}%"
-      column = "(#{column})::text"
-    when 'IN'
-      # empty
-    else
-      fail "Unknown operator: #{operator.inspect}"
+    def query_pattern
+      /([^\ ]*)\ *([:=])\ *("[^"\:]*"|[^\:\ \=]*)/
     end
-    where([column, operator, '(?)'].join(' '), term)
-  end
 
+    def retrieve_users(query, exact_match = false)
+      if exact_match
+        User.where(name: query).each_with_object([]) {|u,out| out << u.id }
+      else
+        User.where(
+          "name ILIKE ?",
+          "%#{query.gsub( "\\", "\\\\\\")}%")
+        .each_with_object([]) {|u,out| out << u.id }
+      end
+    end
 
-  def parts_groups
-    matrix.group_by do |t|
-      t.config_vars_hash['PART'] || t.config_vars_hash['Part']
+    # maps fragment of old state given to travis states
+    def determine_states(query, exact_match = false)
+      states_map = FinalAPI::V1::Http::DDTF_Build::BUILD_STATE2API_V1STATUS
+      states_map.reject { |k,v| k == '' }.each_with_object([]) do |(travis_state, ddtf_state), out|
+        if exact_match
+          out << travis_state if ddtf_state.downcase == query.downcase
+        else
+          out << travis_state if ddtf_state.downcase.include? query.downcase
+        end
+      end.compact
     end
-  end
 
-  # set mandatory properties
-  # this is temporary solution for invalid data in DB
-  # ...and just for development phase
-  def sanitize
-    self.repository ||= Repository.new
-    self.owner ||= User.new
-    self.request ||= Request.new
-    self
+    def retrieve_filter(key, operator, value)
+      exact_match = (operator == '=')
+      case key
+      when 'owner_id', 'stopped_by_id'
+        { key.to_sym => retrieve_users(value, exact_match) }
+      when 'state'
+        { key.to_sym => determine_states(value, exact_match) }
+      else
+        escaped_value = value.gsub(/[%_]/) { |str| "[#{str}]" }
```

Comment:
Nice one, but it seems that:
- this prevents from searching things with underscore on purpose, such as: DDTF_
- there is still problem with semicolon: DDTF;whatever -> whatever gets ignored